### PR TITLE
[chore] deduplicate config types

### PR DESCRIPTION
**Pull request by ignatiusmb**
It's getting harder to track changes that happens to the config options, this will make it so that we derive `ValidatedConfig` from `Config` and only need to change stuff in one place.

### CODE REVIEW

**Comment by benmccann** [comment]
looks like there's a merge conflict

File: packages/kit/types/helper.d.ts [hunk 1]
```diff
@@ -31,3 +31,10 @@ export type InferValue<T, Key extends keyof T, Default> = T extends Record<Key,
 	: Default;
 export type MaybePromise<T> = T | Promise<T>;
 export type Rec<T = any> = Record<string, T>;
+export type RecursiveRequired<T> = {
```
Comment (benmccann) [inline packages/kit/types/helper.d.ts:34]: I can't understand what this is doing. It probably deserves a code comment at least

As an alternative, should we replace `Config` with `Partial<ValidatedConfig>`?

Comment (ignatiusmb) [reply packages/kit/types/helper.d.ts:34]: This is a recursive implementation of [Required](https://www.typescriptlang.org/docs/handbook/utility-types.html#requiredtype), with a few modifications to handle some of our differences between user config and validated config. I'll add a code comment for that one.

Replacing with `Partial<ValidatedConfig>` as an alternative would still need a recursive implementation of `Partial` as it only sets properties one-level deep, `ValidatedConfig` also has `kit: { files: { setup: string } }` which is only used after validation and not in the user config.

File: packages/kit/types/helper.d.ts [hunk 2]
```diff
@@ -31,3 +31,14 @@ export type InferValue<T, Key extends keyof T, Default> = T extends Record<Key,
...
+	// Recursive implementation of Required utility type.
```
Comment (benmccann) [inline packages/kit/types/helper.d.ts:35]:
```suggestion
	// Recursive implementation of TypeScript's Required utility type.
```

**Pull request merged**